improver/lapse_rate.py

```diff
@@ -372,81 +295,65 @@ def _generate_lapse_rate_array(
         Calculate lapse rates and apply filters
 
         Args:
-            temperature_data (numpy.ndarray)
+            temperature_data (numpy.ndarray):
                 2D array (single realization) of temperature data, in Kelvin
-            orography_data (numpy.ndarray)
+            orography_data (numpy.ndarray):
                 2D array of orographies, in metres
-            land_sea_mask_data (numpy.ndarray)
+            land_sea_mask_data (numpy.ndarray):
                 2D land-sea mask
 
         Returns:
-            numpy.ndarray
+            numpy.ndarray:
                 Lapse rate values
         """
-        # Fill sea points with NaN values. Can't use Numpy mask since not
-        # recognised by "generic_filter" function.
+        # Fill sea points with NaN values.
         temperature_data = np.where(land_sea_mask_data, temperature_data, np.nan)
 
-        # Generate data neighbourhoods on which to calculate lapse rates
-        # pylint: disable=unsubscriptable-object
-        dataarray_size = temperature_data.shape[0] * temperature_data.shape[1]
+        lapse_rate_array = []
+        # Pads the data with nans and generates windows representing
+        # a neighbourhood for each point.
 
-        temp_nbhoods = np.zeros(
-            (dataarray_size, self.nbhoodarray_size), dtype=np.float32
-        )
-        fnc = SaveNeighbourhood(allbuffers=temp_nbhoods)
-        generic_filter(
-            temperature_data,
-            fnc.filter,
-            size=self.nbhood_size,
-            mode="constant",
-            cval=np.nan,
+        temp_nbhood_window, orog_nbhood_window = self._create_windows(
+            temperature_data, orography_data
         )
 
-        orog_nbhoods = np.zeros(
-            (dataarray_size, self.nbhoodarray_size), dtype=np.float32
-        )
-        fnc = SaveNeighbourhood(allbuffers=orog_nbhoods)
-        generic_filter(
-            orography_data,
-            fnc.filter,
-            size=self.nbhood_size,
-            mode="constant",
-            cval=np.nan,
-        )
+        # Zips together the windows for temperature and orography
+        # then finds the gradient of the surface temperature with
+        # orography height - i.e. lapse rate.
+        axis = (-2, -1)
+        for temp, orog in zip(temp_nbhood_window, orog_nbhood_window):
```

Comment:
Probably not much difference in performance or memory, but we could preallocate the output array and fill it in place:
```python
lapse_rate_array = np.empty_like(temperature_data)
for laps, temp, orog in zip(lapse_rate_array, temp_nbhood_window, orog_nbhood_window):
    ...
    grad = ...
    ...
    laps[...] = grad
```